internal/dbtools/membership_enumeration.go

```diff
@@ -230,6 +230,82 @@ func GetAllGroupMemberships(ctx context.Context, db *sql.DB, shouldPopulateAllMo
 	return enumeratedMemberships, nil
 }
 
+// CheckNewHierarchyWouldCreateCycle ensures that a new hierarchy does not create a loop or cycle in the database
+func CheckNewHierarchyWouldCreateCycle(ctx context.Context, db *sql.DB, parentGroupID, memberGroupID string) (bool, error) {
+	hierarchies := make(map[string][]string)
+
+	hierarchyRows, err := models.GroupHierarchies().All(ctx, db)
+	if err != nil {
+		return false, err
+	}
+
+	for _, row := range hierarchyRows {
+		hierarchies[row.ParentGroupID] = append(hierarchies[row.ParentGroupID], row.MemberGroupID)
+	}
+
+	hierarchies[parentGroupID] = append(hierarchies[parentGroupID], memberGroupID)
```

Comment:
Do we need to check that `hierarchies[parentGroupID]` exists?

Reply:
I don't think so - when it doesn't already exist we get the zero value of `[]string{}` which is fine, then we assign our new value to it and it exists afterwards. Some of the test cases for this function verify this.